### ISSUE TITLE
Can not extend DbSet with own classes that inherit from DbSet

### ISSUE THREAD

**Issue by Seramis** [dotnet/efcore#12422]
https://github.com/aspnet/EntityFrameworkCore/blob/6bb96ccb241c87b44c005e87d330dc5cadd1e6d7/src/EFCore/Internal/DbSetFinder.cs#L39-L40

Because of these conditions i can not extend my own repository classes from DbSet and define my repository classes in DbContext instead of direct DbSet.

The conditions should not directly check if defined property is DbSet, but if it inherits (or is directly) DbSet.

**Comment by ilmax** [comment]
Hi @Seramis, I'm a bit curious about what scenario are you trying to achieve extending dbset.

**Comment by Seramis** [comment]
Hi @ilmax ,
The DbContext is a full-blown Unit of Work implementation and DbSet is a full-blown Repository implementation.

It's always possible to create your own repository on top of DbSet (effectively creating repository on top of repository), but that's not really reasonable. Also EF Core documentation warns you about not creating secondary UoW and Repo implementations on top of EF Core, that has already implemented all of that.

The fix to support extended DbSet-s is also a one-liner.

**Comment by ilmax** [comment]
Yes, I understand the fix is a one line, but since this is not something very common, I'm interested into the use case of creating your own dbset version.

**Comment by Seramis** [comment]
@ilmax ,
Yes, i understand that. I just mentioned it in addition as it seems, it's not really intended to be that strict. (Based on official documents)

For a example, i would like to create UserRepository : DbSet<User>, where i have additinal method ByGender() and the real LINQ conditions are inside that method.
Like an ordinary Repository in that sense.

**Comment by ajcvickers** [comment]
@Seramis DbSet creation and management is not completely trivial. I'm not against being able to use types derived from DbSet, but such functionality should be done in a way that it works consistently with the way DbSet works, including things like the caching which in turn ensures things like .Local work consistently. We would consider a PR for this if it is done well. The PR would need to have tests showing the experience is consistent.

**Comment by smitpatel** [comment]
Related: `ParameterExtractingExpressionVisitor` also utilizes `typeof(DbSet<>)` and may need to adjust for custom DbSet implementation. Though it could also be costly since its query front-end.

**Comment by d4e666** [comment]
what is the final on this? won't fix or under investyigatyion. I would like to see some REASON for getting closed instead of just "closed question". I see a lot of issues just get closed which make perfect sence. Wo we realy need to start poluting codebases with extension methods because of some lazy development team?

**Comment by smitpatel** [comment]
> We would consider a PR for this if it is done well. The PR would need to have tests showing the experience is consistent.

The value vs cost does not favor implementing this for us (given all other competing features). So we are not going to do it. We close the issue we are not planning to fix (ever) since we don't need to track them.

**Comment by TauanMatos** [comment]
> Hi @ilmax ,
> The DbContext is a full-blown Unit of Work implementation and DbSet is a full-blown Repository implementation.
> 
> It's always possible to create your own repository on top of DbSet (effectively creating repository on top of repository), but that's not really reasonable. Also EF Core documentation warns you about not creating secondary UoW and Repo implementations on top of EF Core, that has already implemented all of that.
> 
> The fix to support extended DbSet-s is also a one-liner.

If I need to implement a custom repository, is it a good idea that my custom repository inherits DbSet<TEntity>?

Or, as you said, "The DbContext is a full-blown Unit of Work implementation", is it ok to use DbContext on an api controller?
